### PR TITLE
Add read_only_adapter_v3

### LINKED ISSUE #3872: Support `make_read_only_endpoint` to be used with historical query adapter
https://github.com/microsoft/CCF/blob/1430f4dc48184ad0002176e2bc8557600299ee9f/samples/apps/logging/logging.cpp#L844-L848

I'm assuming the reason why the historical query adapter is written around `EndpointContext` instead of `ReadOnlyEndpointContext` is to give more flexibility in case something has to written to the KV when a historical query was made. Even though a rare use case, it may exist.

From a dev perspective though it seems odd that this means I have to disable the static guarantees that come from using `make_read_only_endpoint`. It would be nice if the adapter somehow supported read-only endpoint functions as well.

### PR DESCRIPTION
**Pull request by achamayou**
Resolve #3872 

- [x] add read_only_adapter_v3
- [x] convert all relevant endpoints to use it

### CODE REVIEW

**Comment by eddyashton** [comment]
I've done the second part of this (converting existing historical endpoints to use `read_only_adapter_v3`). We don't do this for historical queries in JS, because we find out that they're historical too late (ie - every JS request is may-write as far as dispatch is concerned, only later do we decide it was read-only/historical).

File: src/node/historical_queries_adapter.cpp [hunk 1]
```diff
@@ -397,18 +398,19 @@ namespace ccf::historical
     return true;
   }
 
-  ccf::endpoints::EndpointFunction adapter_v3(
-    const HandleHistoricalQuery& f,
+  template <class HQY, class EFN, class CTX, class EXT>
```
Comment (MahatiC) [inline src/node/historical_queries_adapter.cpp:401]: Perhaps add a small comment indicating what these template abbreviations are? Something like: `EFN = EndpointFunction, CTX=EndpointContext, EXT=Extractor` ? (Not sure what `HQY` is)

Comment (MahatiC) [reply src/node/historical_queries_adapter.cpp:401]: Better yet, name the template parameters as `EndpointFunctionT`, `EndpointContextT`, `ExtractorT` etc ending them with letter `T` to indicate they are template parameters but they also the names give context as to what they represent.

Comment (MahatiC) [reply src/node/historical_queries_adapter.cpp:401]: Better yet, name the template parameters as EndpointFunctionT, EndpointContextT, ExtractorT etc ending them with letter T to indicate they are template parameters but they also the names give context as to what they represent.

Comment (eddyashton) [reply src/node/historical_queries_adapter.cpp]: Good suggestion - I've renamed these template parameters for clarity.

**Review by MahatiC — APPROVED** [review]

**Comment by eddyashton** [comment]
Admin-merging to skip unrelated Daily failure

**Pull request merged**